yaml2rst/tree_node.py

```diff
@@ -0,0 +1,239 @@
+#
+# You should have received a copy of the GNU General Public License
+# along with INSPIRE-SCHEMAS; if not, write to the
+# Free Software Foundation, Inc., 59 Temple Place, Suite 330, Boston,
+# MA 02111-1307, USA.
+#
+# In applying this license, CERN does not
+# waive the privileges and immunities granted to it by virtue of its status
+# as an Intergovernmental Organization or submit itself to any jurisdiction.
+
+from __future__ import print_function, absolute_import
+import os
+
+
+class TreeNode:
+    """Basic representation of a recursive tree data structure.
+
+    Every node has a value field of string type, a list of children of the same type, and a reference to
+    its parent node. Furthermore, every node has a level field, which tells the node's height in a
+    hierarchical structure. When a parent node is provided, the level is increased by 1, otherwise the node is
+    considered a root node, having level 0.
+    """
+
+    def __init__(self, val='', parent=None):
+        """
+        Constructor.
+
+        Create a new TreeNode instance holding the given value.
+        It also comes up with an empty children list.
+        If a parent is provided, this node is appended to its children list and its level is given by
+        increasing by 1 the parent's level.
+
+        Args:
+            val (string): the node's value
+            parent (``TreeNode``): the node's parent.
+        """
+        self.value = unicode(val.strip())
+        self.children = []
+        self.parent = parent
+
+        if parent is not None:
+            parent.children.append(self)
+            self.lvl = parent.lvl + 1
+        else:
+            self.lvl = 0
+
+    def is_leaf(self):
+        """
+        Check if the current TreeNode instance is a leaf node.
+
+        Return:
+            bool: True if the node has at least one child, else False
+        """
+        return len(self.children) is 0
+
+    def __str__(self):
+        return self.value
+
+    def __eq__(self, other):
+        """
+        Compares the current TreeNode instance with the given one in input, applying the following recursive
+        definition: two TreeNode nodes are equal if their value is the same and if their children are
+        equals.
+
+        Args:
+            other: the object to compare with.
+
+        Return:
+            bool: True if objects are equals as defined above, else False.
+        """
+        if not isinstance(other, self.__class__):
+            return False
+
+        if self.value == other.value and len(self.children) == len(other.children):
+            for child, other_child in zip(self.children, other.children):
+                if not child == other_child:
+                    return False
+            return True
+        else:
+            return False
+
+    def __ne__(self, other):
+        return not self.__eq__(other)
+
+    @classmethod
+    def dict2tree(cls, dictionary, root_node):
```

Comment:
If I understand correctly, this is building the tree on top of `root_node`, i.e. modifying it. This is not obvious from the docstring.